core/lib/libmbedtls/src/tee_lmd_provider.c

```diff
@@ -583,53 +683,290 @@ crypto_acipher_ecc_shared_secret(struct ecc_keypair *private_key __unused,
+	}
+
+	*ctx_ret = ctx;
+	return TEE_SUCCESS;
 }
 
-void crypto_cipher_copy_state(void *dst_ctx __unused, void *src_ctx __unused,
-			      uint32_t algo __unused)
+void crypto_cipher_free_ctx(void *ctx, uint32_t algo __maybe_unused)
 {
+	size_t ctx_size __maybe_unused;
+
+	/*
+	 * Check that it's a supported algo, or crypto_cipher_alloc_ctx()
+	 * could never have succeded above.
+	 */
+	assert(!cipher_get_ctx_size(algo, &ctx_size));
+	mbedtls_cipher_free(ctx);
+	free(ctx);
 }
 
-TEE_Result crypto_cipher_init(void *ctx __unused, uint32_t algo __unused,
-			      TEE_OperationMode mode __unused,
-			      const uint8_t *key1 __unused,
-			      size_t key1_len __unused,
-			      const uint8_t *key2 __unused,
-			      size_t key2_len __unused,
-			      const uint8_t *iv __unused,
-			      size_t iv_len __unused)
+void crypto_cipher_copy_state(void *dst_ctx, void *src_ctx,
+				uint32_t algo __maybe_unused)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	mbedtls_cipher_clone(dst_ctx, src_ctx);
 }
 
-TEE_Result crypto_cipher_update(void *ctx __unused, uint32_t algo __unused,
-				TEE_OperationMode mode __unused,
-				bool last_block __unused,
-				const uint8_t *data __unused,
-				size_t len __unused, uint8_t *dst __unused)
+TEE_Result crypto_cipher_init(void *ctx, uint32_t algo,
+			      TEE_OperationMode mode,
+			      const uint8_t *key1, size_t key1_len,
+			      const uint8_t *key2 __maybe_unused,
+			      size_t key2_len __maybe_unused,
+			      const uint8_t *iv __maybe_unused,
+			      size_t iv_len __maybe_unused)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	const mbedtls_cipher_info_t *cipher_info = NULL;
+	int lmd_res;
+
+	if (!ctx)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	cipher_info = tee_algo_to_mbedtls_cipher_info(algo, key1_len * 8);
+	if (cipher_info == NULL)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	lmd_res = mbedtls_cipher_setup_info(ctx, cipher_info);
+	if (lmd_res != 0) {
+		EMSG("setup info failed, res is 0x%x\n", -lmd_res);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	lmd_res = mbedtls_cipher_setkey(ctx, key1, key1_len * 8,
+				mode == TEE_MODE_ENCRYPT ?
+				MBEDTLS_ENCRYPT : MBEDTLS_DECRYPT);
+	if (lmd_res != 0) {
+		EMSG("setkey failed, res is 0x%x\n", -lmd_res);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	if (iv != NULL) {
+		lmd_res = mbedtls_cipher_set_iv(ctx, iv, iv_len);
+		if (lmd_res != 0) {
+			EMSG("set iv failed, res is 0x%x\n", -lmd_res);
+			return TEE_ERROR_BAD_STATE;
+		}
+	}
+
+	lmd_res = mbedtls_cipher_reset(ctx);
+	if (lmd_res != 0) {
+		EMSG("mbedtls_cipher_reset failed, res is 0x%x\n", -lmd_res);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	return TEE_SUCCESS;
 }
 
-void crypto_cipher_final(void *ctx __unused, uint32_t algo __unused)
+TEE_Result crypto_cipher_update(void *ctx, uint32_t algo,
+				TEE_OperationMode mode __maybe_unused,
+				bool last_block __maybe_unused,
+				const uint8_t *data, size_t len, uint8_t *dst)
 {
+	int lmd_res;
+	size_t olen;
+	size_t finish_olen;
+#if defined(CFG_CRYPTO_ECB)
+	size_t blk_size;
```

Comment:
Could write this as `size_t blk_size __maybe_unused` and drop the `if defined`.

Reply:
Yes, it will be better.